### DISCUSSION TITLE
Device unreachable after mqtt reconnect fails

### DISCUSSION THREAD

**Discussion by reimerp** [arendst/Tasmota#16140] [Q&A] [answered]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._

A very strange problem occurs on an ESP32-U4WDH rev.3 based device -- possible even not tasmota is the root cause, but some esp32 libraries / TCP stack / or even the mqtt server (mosquitto)
After a while, I noticed some missing MQTT events from it, and after digging into this (using syslog 2) I noticed that the device lost mqtt connection. "Connect failed to broker, rc -5"
This seems to be a DNS related problem, so to rule that out I used IP addresses instead of names in configuration for
MqttHost / NtpServer1 / LogHost.
The problem occures again, but the error changes to -2, indicating a MQTT connection problem.
Having syslog in place, I find this is still functioning ang logging unsuccessfull mqtt attempts, as well as other results from that device (bluetooth beacons).
The mqtt server (mosquitto version 2.0.11, configured to nearly default) is running fine.
The IP Stack on this device sems not reacting on icmp (ping) messages (or with ~90% packet loss)  while this happens.
WLAN is running fine, and the device is showing best signal when its operationable.
I fired up an tcpdump on the server to capture the mqtt messages, I can see that the device sends regulary MQTT ping messages every 30 seconds which are responded from the server.
From time to time the server is re-sending packages and this seems to completely confuses the device. I don't know why this happens, the resent should occur when no ACK is received, but there are only ~500ms between the resent. The device has only ACKS previous packages, not the duplicate ones and then there is this mqtt lost connection & trying, and maybe the device recoveres, but I have also seen cases where it was completely lost for hours (mqtt in this case) -- syslog seems operational.
The device itself is unreachable, tried ping or reach tcp port 80, so I cannot even issue a reset command.
I have dozens of esp8266 based devices running fine, so this behavior seems esp32 related (or only on the solo1) -- maybe buried inside the TCP/IP stack.
I also dumped a whole day of mqtt traffic for other devices and found not a single duplicate transaction, also no increasing in MqttCount.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [ ] Device used (e.g., Sonoff Basic): Shelly Plus 1 ()
- [x] Tasmota binary firmware version number used: 12.0.2 (release but also checked development branch)
  - [x] Pre-compiled
  - [x] Self-compiled
- [x] Flashing tools used: esptool
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_
Just wait....

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
the device shoud ignore duplicate packages, and ack accordingly

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by reimerp** [comment]
I have read https://github.com/arendst/Tasmota/discussions/10262
with some hints but nothing really helped (yet)

Normally the mqtt Request is responded by my mosqutto server in 0.3-0-5 ms
This packet is than acked by the shelly after 13ms quickest, bur normally after 160ms up to 300ms
If there is no ACK the server  does the TCP Retransmission (mqtt Response) again after ~400ms

I saw this retransmission occasionally on other devices, but there seems no packet loss but only a timeout, so this leads to a duplicate ACK and everything goes back to nomal.

I will do some traces again regarding the ARP traffic, see link above

**Comment by barbudor** [comment]
As you haven't provided the information required by the issue template I must ask : what binary are you using ?
Shelly Plus 1 requires [tasmota32solo1.bin](http://ota.tasmota.com/tasmota32/tasmota32solo1.bin)

You definitively seems to have some WIfi problems. First Tasmota cannot resolve DNS, then when you switch to static you can't connect to MQTT. But suprising that you can get logs through syslog
Have you tried to increase the log level to 3 or 4 ? It may give you more hints on what's Tasmoa is up to which make its so busy

**Comment by reimerp** [comment]
Yes, I used tasmota32solo1.bin first release version, than selft-build from development branch.
The shelly reports 
"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,
"Wifi": "RSSI":100,"Signal":-35,"LinkCount":1,"Downtime":"0T00:00:05"
"MqttCount":27
It has one of the best WIFI connection from all my devices but the highest MqttCount, others count +1 in 2-3 days....
Currently I am running a full TCPDump and to not disturb that I set syslog to 0

**Comment by barbudor** [comment]
MqttCount clearly shows that it has problem with MQTT connections
So the problem is not the Wifi
Are you sure to use unique MqttClient for each of your devices ?
Using the same MqttClient on 2 different devices would lead to continuous disconnection/reconnections as Mosquitto want only 1 connection per MqttClient
(I believe you know the difference between the MttqClient and MqttUser/MqttPassword)

**Comment by reimerp** [comment]
Yes sure, MqttClient ist kept on default:"DVES_%06X" which is build from MAC address. A duplicate client Id would surely harm the second device and I would expect a more obvious fault. My first dumps shows relative clearly, that the device is not sending TCP ACKs in time and this leads to retransmission, wrong ACKs and the device thinking of Mqtt connection loss. To recover from this it can take ~120s up to several minutes, see dumps below. Unfortunately I captured only mqtt traffik, so I couldn't check if e.g. ARP/DHCP is doing nasty things in between -- i will repeat.
![Bildschirmfoto vom 2022-08-02 23-39-09](https://user-images.githubusercontent.com/5211658/182478606-82c35959-c2a4-45d1-9bfc-1458dd652531.png)

**Reply by barbudor** [reply]
No more ideas
May be something very specific with solo1 as I don't have any of those

**Reply by sfromis** [reply]
I have several solo1 devices without experiencing any MQTT or Wifi trouble. Of course, networking-related issues may be AP-dependent.

**Comment by reimerp** [comment]
After running with few Retransmissions for several hours, but no MqttCount increase, there was a majour outtage of the device for more than 100 minutes. WiFi / Mosquitto was stable all the time. Latest Load & Wifi was ok from latest mqtt messages. It started with a Retransmission from server, Duplicate ACK and FIN on the connection after ~80s. After that the device gets completely lost. There are ARP Requests from the server which are succesfully answered, but the Device needed ~ 10 s for answer ( normal is <300ms) but for sure this seems not the root cause.
I do have a tcpdump here, if smbd likes, but
I have no more ideas.

**Comment by bongibong** [comment]
im running a similar issue..
installed version Tasmota 12.0.2 on several of my sonoff devices.
all devices with that version without any configuration change, reports mqtt connection issue:
MQT: Connect failed to raspberrypi.local:1883, rc -5. Retry in 120 sec
other devices that still on version 11, works fine.
when i set the device mqtt server to the my raspberrypi server IP instead of raspberrypi.local it works fine!
pinging raspberrypi.local from my laptop works.

STATUS 0:
CMD: STATUS 0
18:42:11.889 RSL: STATUS = {"Status":{"Module":29,"DeviceName":"OutLight2","FriendlyName":["OutLight2","OutLight2"],"Topic":"OutLight2","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
18:42:11.896 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T00:25:53","StartupUTC":"2022-08-03T17:16:18","Sleep":50,"CfgHolder":4617,"BootCount":45,"BCResetTime":"2021-04-03T14:57:21","SaveCount":1329,"SaveAddress":"FA000"}}
18:42:11.903 RSL: STATUS2 = {"StatusFWR":{"Version":"12.0.2(tasmota)","BuildDateTime":"2022-06-20T12:41:03","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8285","CR":"404/699"}}
18:42:11.908 RSL: STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["bongi",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A000000000000","00000080","00006000","00004000","00000000"]}}
18:42:11.917 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":619,"Free":384,"Heap":26,"ProgramFlashSize":1024,"FlashSize":1024,"FlashChipId":"144051","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8F9AC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","14000020","00000000"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45,62","Sensors":"1,2,3,4,5,6"}}
18:42:11.924 RSL: STATUS5 = {"StatusNET":{"Hostname":"OutLight2-0287","IPAddress":"10.100.102.3","Gateway":"10.100.102.1","Subnetmask":"255.255.255.0","DNSServer1":"10.100.102.1","DNSServer2":"1.1.1.1","Mac":"24:A1:60:xx:xx:xx","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
18:42:11.927 RSL: STATUS6 = {"StatusMQT":{"MqttHost":"raspberrypi.local","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_10811F","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
18:42:11.933 RSL: STATUS7 = {"StatusTIM":{"UTC":"2022-08-03T17:42:11","Local":"2022-08-03T18:42:11","StartDST":"2022-03-27T02:00:00","EndDST":"2022-10-30T03:00:00","Timezone":"+01:00","Sunrise":"05:26","Sunset":"20:25"}}
18:42:11.938 RSL: STATUS10 = {"StatusSNS":{"Time":"2022-08-03T18:42:11"}}
18:42:11.943 RSL: STATUS11 = {"StatusSTS":{"Time":"2022-08-03T18:42:11","Uptime":"0T00:25:53","UptimeSec":1553,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"bongi","BSSId":"10:13:31:zz:zz:zz","Channel":11,"Mode":"11n","RSSI":66,"Signal":-67,"LinkCount":1,"Downtime":"0T00:00:06"}}}


weblog 4:
00:00:00.001 HDW: ESP8285
00:00:00.050 CFG: Loaded from flash at F9, Count 1330
00:00:00.055 QPC: Count 1
00:00:00.057 CFG: CR 404/699, Busy 0
00:00:00.062 ROT: Mode 1
00:00:00.063 SRC: Restart
00:00:00.065 Project tasmota - OutLight2 Version 12.0.2(tasmota)-2_7_4_9(2022-06-20T12:41:03)
00:00:00.219 WIF: Checking connection...
00:00:00.220 WIF: Attempting connection...
00:00:00.556 WIF: Connecting to AP1 bongi Channel 11 BSSId 10:13:31:23:65:4D in mode 11n as OutLight2-0287...
00:00:01.802 WIF: Checking connection...
00:00:01.803 WIF: Connected
00:00:02.004 HTP: Web server active on OutLight2-0287 with IP address 10.100.102.3
00:00:02.403 WIF: Sending Gratuitous ARP
00:00:02.404 NTP: Sync time...
00:00:02.448 RTC: UTC 2022-08-03T17:44:57, DST 2022-03-27T02:00:00, STD 2022-10-30T03:00:00
18:44:57.000 RTC: Synced by NTP
18:44:58.006 MQT: Attempting connection...
18:44:58.029 DNS: Unable to resolve 'raspberrypi.local'
18:44:58.030 MQT: Connect failed to raspberrypi.local:1883, rc -5. Retry in 10 sec
18:45:01.485 QPC: Reset
18:45:01.946 HTP: Main Menu
18:45:03.223 HTP: Console
18:45:03.384 APP: Boot Count 46
18:45:03.624 CFG: Saved to flash at F8, Count 1331, Bytes 4096
18:45:09.037 MQT: Attempting connection...
18:45:09.065 DNS: Unable to resolve 'raspberrypi.local'
18:45:09.066 MQT: Connect failed to raspberrypi.local:1883, rc -5. Retry in 20 sec
18:45:16.793 WIF: Checking connection...
18:45:30.008 MQT: Attempting connection...
18:45:30.036 DNS: Unable to resolve 'raspberrypi.local'
18:45:30.037 MQT: Connect failed to raspberrypi.local:1883, rc -5. Retry in 30 sec

**Reply by barbudor** [reply]
@bongibong no you are not facing the same issue. You don't have your ESP crashing because of heap exhaust

What you are facing is expecting a feature which is not available by default on Tasmota because it used to create a lot of problem in the past: mDNS (names in xxx.local)
That features works on your PC, fine. But it is not available on Tasmota. If you want to use names instead of IP, configure a true local DNS server and do not rely on mDNS

**Reply by bongibong** [reply]
Thanks @barbudor, how do you explain that on the same switch and on all other sonoff switches i have with the previous tasmota version (11) all worked and works fine ?? nothing has changed on my network except the new tasmota version installation

**Reply by barbudor** [reply]
@bongibong 
Apologies I mistook this thread with another one. There is no ESP crash due to heap exhaust here, but device becoming un available.

There was a change introduced in 12.0.0 regarding DNS management, followed by a fix in 12.0.1. [CHANGELOG](https://github.com/arendst/Tasmota/blob/development/CHANGELOG.md)
We are not supposed to have lost functionality.

However, @reimerp problem seems to be in 2 parts:
1) A DNS problem when using a name for it's MQTT leading to error -5 like yours. However, my understanding from @reimerp is that initial MQTT connection at boot seems ok, but later on MQTT disconnect and problems starts to occur with device being unreachable
2) When using static IP instead of name (removing the DNS from the equation), same behavior with connection after boot ok but followed by lost of connections and MQTT error -2 and device being unreachable

@bongibong to be sure if your problem is the same or not:
Use fixed IP address instead of name
Then answer this questions: Do you face MQTT disconnection leading to -2 errors and device being unreachable ?
If yes, your probem is the same
If no, your problem is strictly a DNS problem. So please open a separated discussion based on "DNS resolution for raspberrypi.local stopped working in 12.0.2"

Thanks

**Reply by bongibong** [reply]
setting static IP works good!!
opened a new discussion here: https://github.com/arendst/Tasmota/discussions/16150

**Comment by reimerp** [comment]
The initial connection and dns resolution works fine, normally for hours. There are rare cases where TCP retransmissions happen, which seems normal when the device is not ACK in time. Normally it revovers. But sometimes not. DNS / MQTT server is not the root cause, its working fine while this happens. As I said, there seems to be huge workload while this happens, this  seems to infect DNS resolution and also Mqtt connection. Unfortunately I cannot dump DNS requests/reponses due to network layout, but to rule this out I switched to static adresses. It could be that DNS request (which is UDP) are answered from my server, but  the Device is not willing to react (blocked by load, deadlocks etc). Since its UDP this is easy, if the device is not able to get the respone, than it should just ask again. Same applies for Syslog Traffic, it's fire & forget. There are packets logged while the device hangs, but no way to check it they are send when the event occured or ms or seconds later, normally this does not interest or is intended to not prioritise this traffic at all. Since the device is not reacting to ICMP request in time or with a huge package loss, or the device is not even willing to answer on HTTP traffic, i would think of load problems or Deadlocks inside the TCP stack...
Anyhow. I have a second identical shelly and I unmounted the problematic one and exchanged them. I also did a clean factrory flash on both with a nearly identical configuration. There are no Rules deployed or no scripts, I'd like to have Timers and a Rule for racting on Switch (Auto Power Off with a RuleTimer) but first of all I have to ensure stability.

**Comment by Jason2866** [comment, marked as answer]
@reimerp Please try latest development solo1 build. The underlying arduino framework and the libs are builded with a very recent IDF 4.4 (devel) version. espressif did changes/fixes in the wifi/lwip stack (closed source). The issue seems is AP dependent.

What brand / type are your APs?

**Reply by reimerp** [reply]
I will re-build and switch development with fresh build and facory flash.
My Router is an AVM Fritzbox 6490, latest Firmware but Freetzed, so I am running a eg. dnsmasq for dns/dhcp.
But I just came across issue #14985 -- so this might be a similar issue buried in the core for solo1, though I never encountered this particular Wifi connect issue..

**Reply by Jason2866** [reply]
ESP32 and ESP32solo1 uses exact the same IDF/Arduino code. It is just one! option different set in `sdkconfig` for building the libs for the Arduino framework. The closed source part for lwip/wifi is 100% the same for ESP32 and ESP32solo1.
If you have a spare ESP32 please flash the solo1 build to. YES solo1 build does run without any issues on a `normal` ESP32. If this device runs without problems there are only two options left. Device is faulty or the ESP32-U4WDH rev.3 does not work correctly with the framework.
2nd Option i doubt since many users flash Tasmota on Shelly devices. 

It is no uncommon that faulty hardware is showing wifi problems. My experience over the time, before digging into a problem i try a second device if it has the same behaviour.

**Reply by reimerp** [reply]
Having switched the devices and upgraded to core 2_0_4_1 from the dev branch, the issue seems gone. I also have the other device upgrade, I do not think its faulty since its also up since a week w/o outages, so for sure this is issue is for sure related to the core.

**Reply by Jason2866** [reply]
Great to hear the issue has gone.
